### PR TITLE
Wgpu 0.20

### PR DESCRIPTION
**Pull request by Elabajaba**
Currently blocked on https://github.com/gfx-rs/wgpu/issues/5774

# Objective

Update to wgpu 0.20

## Solution

Update to wgpu 0.20 and naga_oil 0.14.

## Testing

Tested a few different examples on linux (vulkan, webgl2, webgpu) and windows (dx12 + vulkan) and they worked.

---

## Changelog

- Updated to wgpu 0.20. Note that we don't currently support  wgpu's new pipeline overridable constants, as they don't work on web currently and need some more changes to naga_oil (and are somewhat redundant with naga_oil's shader defs). See wgpu's changelog for more https://github.com/gfx-rs/wgpu/blob/trunk/CHANGELOG.md#v0200-2024-04-28

## Migration Guide

TODO

### CODE REVIEW

**Comment by Elabajaba** [comment]
Technically we should wait for https://github.com/gfx-rs/wgpu/pull/5681 to be backported before landing this, but that's supposed to be merged and released tomorrow (May 29th).

**Review by IceSentry — APPROVED** [review]
I tested a bunch of 2d and 3d examples on windows 10 with a 4080 and everything worked correctly.

Would be nice not having TODOs, but they can be addressed later. We should still ship 0.14 with this

LGTM

**Comment by Elabajaba** [comment]
> I tested a bunch of 2d and 3d examples on windows 10 with a 4080 and everything worked correctly.
> 
> Would be nice not having TODOs, but they can be addressed later. We should still ship 0.14 with this
> 
> LGTM

The Todos shouldn't have any effect anyways, since they're for pipeline override constants which don't work on web currently (and aren't yet supported in naga_oil).

**Comment by SIGSTACKFAULT** [comment]
examples/3d_shapes dies of SIGBUS after ~5s on NixOS Unstable, RX6600.

https://github.com/bevyengine/bevy/assets/14184826/574313b4-3967-430a-8893-f2647a405bc9

with `--features wayland` it brought Plasma down with it.

**Review by JMS55 — APPROVED** [review]
Tested a bunch of 3d examples with VK + DX12, lgtm.

```
2024-05-29T06:50:45.203720Z  INFO bevy_diagnostic::system_information_diagnostics_plugin::internal: SystemInfo { os: "Windows 11 Home", kernel: "22631", cpu: "Intel(R) Core(TM) Ultra 7 155H", core_count: "16", memory: "15.6 GiB" }
2024-05-29T06:50:45.241110Z  INFO bevy_winit::system: Creating new window "App" (Entity { index: 0, generation: 1 })
2024-05-29T06:50:45.572420Z  INFO bevy_render::renderer: AdapterInfo { name: "Intel(R) Arc(TM) Graphics", vendor: 32902, device: 32085, device_type: IntegratedGpu, driver: "", driver_info: "", backend: Dx12 }
```

**Comment by killercup** [comment]
Tested a bunch of examples on an Apple M1 Pro and they all looked good.

**Comment by Henauxg** [comment]
The `render_to_texture` example does not work for me on this branch, the inside cube does not rotate.
It does works on `main`.
Tested on Windows 11 with a 4070 + both Vulkan & Dx12.

**Comment by IQuick143** [comment]
Did testing on Linux (NVIDIA, KDE, XOrg):

The `two_passes` example does not render anything:
image
Also causes major lag to the whole system (UI & Audio froze up for a few seconds) when opening and closing the window.

Equivalent behaviour in the `split_screen` example.

Similar hangs occur when a lot of resizing is done.

Other than that all 3d examples (except meshlets and clearcoat which I haven't tested) work.

**Comment by IQuick143** [comment]
Could you elaborate a bit more on how to do the patching?
I get `warning: Patch `wgpu v0.20.0 (https://github.com/cwfitzgerald/wgpu.git?branch=vk/fix-presentation-synchronization#4cf108c4)` was not used in the crate graph.` after putting
```
[patch.'https://github.com/gfx-rs/wgpu.git']
wgpu = { git = 'https://github.com/cwfitzgerald/wgpu.git', branch = 'vk/fix-presentation-synchronization' }
```
In the root Cargo.toml

**Comment by IQuick143** [comment]
After trying that I get:
"
error: failed to select a version for `wgpu`.
versions that meet the requirements `^0.20` are: 0.20.0

the package `bevy_render` depends on `wgpu`, with features: `naga` but `wgpu` does not have these features.
 It has an optional dependency with that name, but that dependency uses the "dep:" syntax in the features table, so it does not have an implicit feature with that name.


failed to select a version for `wgpu` which could resolve this conflict
"

**Comment by Friz64** [comment]
Damn, I assumed the PR would still be compatible with 0.20. That's https://github.com/gfx-rs/wgpu/pull/5691, which we eventually need to but already can fix on our side. If there's still something breaking after that, a backport PR is required.

**Comment by JMS55** [comment]
Actually I forgot, transmission seemed broken to me.

**Comment by alice-i-cecile** [comment]
Transmission looks fine (or at least, identical to `main`) on Windows 11.

**Comment by alice-i-cecile** [comment]
> Wumpf: re wgpu: I'd strongly recommend waiting for 0.20.1. The still to-be-merged issue is pretty far reaching for all Vulkan users 

- [Wumpf, one of the wgpu maintainers](https://discord.com/channels/691052431525675048/1245744109927272549/1245750832327032922)

**Comment by BD103** [comment]
Is there an issue / PR we can link to that's blocking this? I saw that @Elabajaba ran into [some Vulkan issues](https://discord.com/channels/691052431525675048/1245744109927272549/1245771589379821660) that pushed this to 0.15.

**Comment by Elabajaba** [comment]
> Is there an issue / PR we can link to that's blocking this? I saw that @Elabajaba ran into [some Vulkan issues](https://discord.com/channels/691052431525675048/1245744109927272549/1245771589379821660) that pushed this to 0.15.

I tracked down what's causing the crash (using multiple command encoders) and made a wgpu issue for it https://github.com/gfx-rs/wgpu/issues/5774.

**Comment by cwfitzgerald** [comment]
v0.20.1 is out now, thanks @Elabajaba for pushing this through!

**Comment by JMS55** [comment]
We should check if this is fixed: https://github.com/bevyengine/bevy/blob/2cffd14923c9e217dc98881af6c8926fe4dc65ea/crates/bevy_render/src/renderer/mod.rs#L372

**Comment by mockersf** [comment]
no issue in CI after the wgpu 0.20.1 update on the examples 🎉

File: crates/bevy_color/Cargo.toml
Comment (BD103) [inline crates/bevy_color/Cargo.toml:20]:
```suggestion
wgpu-types = { version = "0.20.1", default-features = false, optional = true }
```

If 0.20.0 was so problematic, we should prevent it from being pulled in by making 0.20.1 the minimum version.

File: crates/bevy_render/Cargo.toml
Comment (BD103) [inline crates/bevy_render/Cargo.toml:73]:
```suggestion
wgpu = { version = "0.20.1", default-features = false, features = [
```

**Review by BD103 — CHANGES_REQUESTED** [review]

**Comment by alice-i-cecile** [comment]
I'm planning to merge this in right at the start of the 0.15 cycle.

**Comment by xangelix** [comment]
> I'm planning to merge this in right at the start of the 0.15 cycle.

Is it possible we could get this in 0.14 and address any stability issues, if they happen, in further 0.14.x releases?
The upcoming release of `egui` will be on `wgpu` 0.20 and I'd like to keep the `wgpu` versions in sync if possible. https://github.com/emilk/egui/pull/4560

**Comment by alice-i-cecile** [comment]
I'll check with the rendering SMEs and other maintainers :)

**Comment by alice-i-cecile** [comment]
Great; there's agreement to ship this in 0.14. Let me do a quick pass then let's get this merged.

**Comment by alice-i-cecile** [comment]
Diff looks reasonable, and it seems like the other issues are now resolved. Merging!

**Comment by xangelix** [comment]
> Great; there's agreement to ship this in 0.14. Let me do a quick pass then let's get this merged.

Awesome, thanks so much!

**Review by mockersf — APPROVED** [review]

**Pull request merged**
